src/transformStateWithClones.js

```diff
@@ -7,7 +7,54 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const propertiesOfActios = [];
+  const stateCopy = JSON.parse(JSON.stringify(state));
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        addProperties(stateCopy, action.extraData);
+        addToSolutionArray(propertiesOfActios, stateCopy);
+        break;
+
+      case 'removeProperties':
+        removeProperties(stateCopy, action.keysToRemove);
+        addToSolutionArray(propertiesOfActios, stateCopy);
+        break;
+
+      case 'clear':
+        if (Object.keys(stateCopy).length === 0) {
+          addToSolutionArray(propertiesOfActios, stateCopy);
+          break;
+        } else {
+          clearProperties(stateCopy);
+          addToSolutionArray(propertiesOfActios, stateCopy);
+        }
```

Comment:
The `if` statement inside the 'clear' case is unnecessary. The `clearProperties` function will handle an already empty state correctly, so you can call `clearProperties` and `addToSolutionArray` directly without the `if` check.